### PR TITLE
[O2B-1241] Add filtering by Alice L3/dipole Current

### PR DESCRIPTION
**Pull request by xsalonx**
#### I have a JIRA ticket
- [x] branch and/or PR name(s) include(s) JIRA ID
- [x] issue has "Fix version" assigned
- [ ] issue "Status" is set to "In review"
- [x] PR labels are selected

Notable changes for users:
- NA

Notable changes for developers:
- Added runs' filtering by aliceL3Current and aliceDipoleCurrent

Changes made to the database:
- NA

### CODE REVIEW

File: lib/public/utilities/fetch/buildUrl.js [hunk 1]
```diff
@@ -74,22 +73,24 @@ export const buildUrl = (baseURL, parameters) => {
...
-            buildParameterFromNestedKeys(parameters, nestedKeys, value, null);
+            const parameterElements = formattedParameter.split(/\[|\]\[|\]/);
+            const nestedKeys = parameterElements.slice(0, parameterElements.length - 1);
+            const [lastKey, value] = parameterElements[parameterElements.length - 1].split('=');
+
+            if (lastKey === '') {
+                buildParameterFromNestedKeys(parameters, nestedKeys, value, null);
+            } else { // List nestedKeys is empty
+                buildParameterFromNestedKeys(parameters, [lastKey], value, null);
+            }
```
Comment (martinboulais) [inline lib/public/utilities/fetch/buildUrl.js:88]: Why this change?

File: lib/public/utilities/fetch/buildUrl.js [hunk 2]
```diff
@@ -74,22 +73,24 @@ export const buildUrl = (baseURL, parameters) => {
         }
     };
 
+    const [origin, existingParameters] = baseURL.split('?');
```
Comment (martinboulais) [inline lib/public/utilities/fetch/buildUrl.js:76]: It is not origin. For example `http://example.com/route` has `http://example.com` as origin, not the `/route`

File: lib/public/utilities/fetch/buildUrl.js [hunk 3]
```diff
@@ -98,6 +99,7 @@ export const buildUrl = (baseURL, parameters) => {
      * @return {void}
      */
     const formatAndPushQueryParameter = (key, value) => {
+        key = key.replace(/=/g, '%3D');
```
Comment (martinboulais) [inline lib/public/utilities/fetch/buildUrl.js:102]: Why this?

File: lib/public/views/Runs/ActiveColumns/runsActiveColumns.js [hunk 4]
```diff
@@ -492,4 +492,11 @@ export const runsActiveColumns = {
         format: (_, run) => formatAliceCurrent(run.aliceDipolePolarity, run.aliceDipoleCurrent),
         profiles: ['runsPerLhcPeriod', 'runsPerDataPass', 'runsPerSimulationPass'],
     },
+    aliceL3AndDipoleCurrent: {
+        name: 'L3 / Dipole',
+        visible: false,
+        filter: ({ aliceL3AndDipoleCurrentFilter }) => selectionDropdown(aliceL3AndDipoleCurrentFilter, { selectorPrefix: 'l3-dipole-current' }),
+        filterTooltip: 'Combinations of L3 and dipole currents of physics runs rounded to kiloamperes',
```
Comment (martinboulais) [inline lib/public/views/Runs/ActiveColumns/runsActiveColumns.js:499]: I think that the name is explicit enough. Were you asked for this tooltip?

File: lib/server/controllers/runs.controller.js [hunk 5]
```diff
@@ -308,6 +311,26 @@ const listReasonTypes = async (_request, response, _next) => {
...
+// eslint-disable-next-line valid-jsdoc
+/**
+ * Retrive distinct combination of levels of alice L3 and dipole current rounded to kilo amperes
+ */
+const listAllAliceL3AndDipoleLevelsForPhysicsRuns = async (req, res) => {
+    const validatedDTO = await dtoValidator(
+        DtoFactory.queryOnly(),
+        req,
+        res,
+    );
```
Comment (martinboulais) [inline lib/server/controllers/runs.controller.js:323]: You don't need this

File: lib/usecases/run/GetAllRunsUseCase.js [hunk 6]
```diff
@@ -143,6 +145,28 @@ class GetAllRunsUseCase {
             if (triggerValues) {
                 filteringQueryBuilder.where('triggerValue').oneOf(...triggerValues);
             }
+            if (aliceL3Current) {
+                /**
+                 * The alice l3 current computed from its absolute value and polarity
+                 * @param {function} literal function to create an object representing a database literal expression
+                 * @return {Object} the object representing the column
+                 */
+                const computedColumn = ({ literal }) => literal("alice_l3_current * (1 - 2 * (`alice_l3_polarity` = 'NEGATIVE'))");
```
Comment (martinboulais) [inline lib/usecases/run/GetAllRunsUseCase.js:154]: Why is this? This do not seem coherent with the pre-defined versions? Also, `IF` is way more readable

File: lib/server/services/run/RunService.js [hunk 7]
```diff
@@ -275,6 +277,28 @@ class RunService {
...
+        const queryBuilder = dataSource.createQueryBuilder()
+            .set('attributes', (sequelize) => [
+                [sequelize.literal('DISTINCT ROUND(IF(alice_l3_polarity = \'NEGATIVE\', -1, 1) * alice_l3_current / 1000)'), 'l3Level'],
+                [sequelize.literal('ROUND(IF(alice_dipole_polarity = \'NEGATIVE\', -1, 1) * alice_dipole_current / 1000)'), 'dipoleLevel'],
+            ])
+            .where('aliceL3Current').not().is(null)
+            .where('aliceL3Polarity').not().is(null)
+            .where('aliceDipoleCurrent').not().is(null)
+            .where('aliceDipolePolarity').not().is(null)
+            .where('definition').is(RunDefinition.Physics);
```
Comment (martinboulais) [inline lib/server/services/run/RunService.js:296]: I think we also want to filter on good.
We want all the combinations, you should probably use the `group by`

As reminder, the request that gave me good results: 
```sql
SELECT DISTINCT CONCAT(
                    IF(alice_l3_polarity = 'NEGATIVE' AND ROUND(alice_l3_current / 1000) <> 0, '-', ''),
                    ROUND(alice_l3_current / 1000),
                    'kA/',
                    IF(alice_dipole_polarity = 'NEGATIVE' AND ROUND(alice_dipole_current / 1000) <> 0, '-', ''),
                    ROUND(alice_dipole_current / 1000),
                    'kA'
                )
FROM runs
WHERE alice_l3_current IS NOT NULL
  AND alice_l3_polarity IS NOT NULL
  AND alice_dipole_current IS NOT NULL
  AND alice_dipole_polarity IS NOT NULL
  AND definition = 'PHYSICS'
  AND run_quality = 'good'
ORDER BY alice_l3_polarity, alice_l3_current, alice_dipole_current, alice_dipole_current;
```

File: lib/domain/dtos/filters/NumericalComparisonDto.js [hunk 8]
```diff
@@ -13,7 +13,12 @@
...
-    operator: Joi.string().valid('<', '<=', '=', '>=', '>'),
+const operators = ['<', '<=', '=', '>=', '>'];
+
+exports.NonNegativeIntegerComparisonDto = Joi.object({
+    operator: Joi.string().valid(...operators),
     limit: Joi.number().integer().min(0),
 });
+
+exports.OperatorToFloatLimitComparisonDto = Joi.object(Object
+    .fromEntries(operators.map((operator) => [operator, Joi.number()])));
```
Comment (martinboulais) [inline lib/domain/dtos/filters/NumericalComparisonDto.js:24]: Does this belong to this PR?

File: lib/public/components/Filters/RunsFilter/AliceL3AndDipoleFilteringModel.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+                        aliceL3Current: {
+                            [l3Level < 0 ? '>' : '>=']: l3Level * 1000 - 500,
+                            [l3Level < 0 ? '<=' : '<']: l3Level * 1000 + 500,
+                        },
+                        aliceDipoleCurrent: {
+                            [dipoleLevel < 0 ? '>' : '>=']: dipoleLevel * 1000 - 500,
+                            [dipoleLevel < 0 ? '<=' : '<']: dipoleLevel * 1000 + 500,
+                        },
+                    },
+                })));
```
Comment (martinboulais) [inline lib/public/components/Filters/RunsFilter/AliceL3AndDipoleFilteringModel.js:50]: What is all that? For me, you should simply have a dropdown with pre-defined options: 
image
And on select you can split the value or whatever

**Review by martinboulais — CHANGES_REQUESTED** [review]

File: lib/public/utilities/fetch/buildUrl.js (hunk 1 above)
Comment (xsalonx) [reply lib/public/utilities/fetch/buildUrl.js:88]: because, now you can put numeric operators as part of nested keys

File: lib/public/utilities/fetch/buildUrl.js (hunk 2 above)
Comment (xsalonx) [reply lib/public/utilities/fetch/buildUrl.js:76]: fixed

File: lib/public/utilities/fetch/buildUrl.js (hunk 3 above)
Comment (xsalonx) [reply lib/public/utilities/fetch/buildUrl.js:102]: to use '=' as numeric operator and as part of nested keys

File: lib/public/views/Runs/ActiveColumns/runsActiveColumns.js (hunk 4 above)
Comment (xsalonx) [reply lib/public/views/Runs/ActiveColumns/runsActiveColumns.js:499]: I would not say reason of having dropdown with a few options there is obvious. This brings glimpse of information why and what's that

File: lib/server/controllers/runs.controller.js (hunk 5 above)
Comment (xsalonx) [reply lib/server/controllers/runs.controller.js:323]: it's agains other controllers in which excessive parameters cause request failure

File: lib/usecases/run/GetAllRunsUseCase.js (hunk 6 above)
Comment (xsalonx) [reply lib/usecases/run/GetAllRunsUseCase.js:154]: applied IF, what are predefined filtering options on fronted has nothing to do with filtering here , alice L3/dipole current are numeric properties so should be filterable as numbers, not as categorical properties

File: lib/server/services/run/RunService.js (hunk 7 above)
Comment (xsalonx) [reply lib/server/services/run/RunService.js:296]: not on good, as data passes can have bad runs. There are all combinations, `group by` doesn't work here there is `distinct`

File: lib/domain/dtos/filters/NumericalComparisonDto.js (hunk 8 above)
Comment (xsalonx) [reply lib/domain/dtos/filters/NumericalComparisonDto.js:24]: Yes, I need `OperatorToFloatLimitComparisonDto` and previous `NumericComparisonDto` had to generic name, so could be mistaken with new one

File: lib/public/components/Filters/RunsFilter/AliceL3AndDipoleFilteringModel.js (hunk 9 above)
Comment (xsalonx) [reply lib/public/components/Filters/RunsFilter/AliceL3AndDipoleFilteringModel.js:50]: And then I would need to create the same objects of filtering parameters, but extract levels from string additionally

File: lib/public/views/Runs/ActiveColumns/runsActiveColumns.js (hunk 4 above)
Comment (martinboulais) [reply lib/public/views/Runs/ActiveColumns/runsActiveColumns.js:499]: I disagree, but ok

**Review by martinboulais — APPROVED** [review]
